.github/workflows/pr_cluster_workflow.yml

```diff
@@ -0,0 +1,174 @@
+    steps:
+    - uses: actions/checkout@v4
+
+    - uses: actions/setup-node@v4
+      with:
+        node-version: 20.14.0
+
+    - name: Setup Terraform Environment
+      uses: hashicorp/setup-terraform@v3
+      with:
+        cli_config_credentials_token: ${{ secrets.TF_API_TOKEN }}
+        terraform_version: 1.7.5
+
+    - name: Configure Terraform Cache
+      run: |
+        echo "TF_PLUGIN_CACHE_DIR=$HOME/.terraform.d/plugin-cache" >> "$GITHUB_ENV"
+
+    - name: Initializing Terraform
+      id: init
+      run: terraform init -upgrade
+      env:
+        TF_WORKSPACE: "default-ws"
+
+    - name: Extract values and set output variables
+      id: output_variables
+      run: |
+        if [[ ${{ github.event.pull_request.draft }} == true ]]; then
+          clusters=$(yq e 'keys | .[] | select(. == "${{ env.TF_VAR_cluster_name }}")' ${GITHUB_WORKSPACE}/config.yaml | jq -R -s -c 'split("\n")[:-1]')
+        else
+          clusters=$(yq e 'keys | .[] | select(. != "${{ env.TF_VAR_cluster_name }}")' ${GITHUB_WORKSPACE}/config.yaml | jq -R -s -c 'split("\n")[:-1]')
+        fi
+        echo "clusters=$clusters" >> $GITHUB_OUTPUT
+
+  update_cluster:
+    needs: setup_cluster
+    if: ${{ needs.setup_cluster.outputs.clusters != '[]' && github.event.pull_request.draft == true }}
+    runs-on: self-hosted
+    permissions:
+      contents: read
+      pull-requests: write
+    defaults:
+      run:
+        working-directory: "terraform"
+    steps:
+    - uses: actions/setup-node@v4
+      with:
+        node-version: 20.14.0
+
+    - name: Infrastructure updates
+      uses: ./.github/actions/apply_terraform_composite_action
+      with:
+        workspace: ${{ env.TF_VAR_cluster_name }}-infrastructure
+        plan_output: tfplan-${{ env.TF_VAR_cluster_name }}-infrastructure
+        working_directory: "terraform/infrastructure"
+        tf_api_token: ${{ secrets.TF_API_TOKEN }}
+
+    - name: Cluster updates
+      uses: ./.github/actions/apply_terraform_composite_action
+      with:
+        workspace: ${{ env.TF_VAR_cluster_name }}-cluster
+        plan_output: tfplan-${{ env.TF_VAR_cluster_name }}-cluster
+        working_directory: "terraform/cluster"
+        tf_api_token: ${{ secrets.TF_API_TOKEN }}
+
+    - name: Generating kube-config
+      uses: actions/upload-artifact@v4
+      with:
+        name: kube-config-${{ env.TF_VAR_cluster_name }}
+        path: /opt/kubeconfig/${{ env.TF_VAR_cluster_name }}
+        compression-level: 0
+
+    - name: Generating Markdown
+      run: |
+        echo "### turnk8s" >> $GITHUB_STEP_SUMMARY
+        echo "" >> $GITHUB_STEP_SUMMARY
+        echo "Push your Kubernetes service manifests to [GitHub URL](https://github.com/infraheads/${{ env.TF_VAR_cluster_name }}) to get them deployed on the cluster. :star_struck:" >> $GITHUB_STEP_SUMMARY
+        echo "Use `kubeconfig` file(s) to connect to the cluster, see the link above :point_up:" >> $GITHUB_STEP_SUMMARY
+
```

Comment:
We should not roll back automatically. We can have a separate rollback workflow or just re-run the existing workflow. The second option for terraform seems more logical, as it promises idempotence. We need to think about the rollback. For example, what are the scenarios in which the rollback may be required? If We need to roll back to another version of the cluster(HCL code), a simple re-run should be enough, as Terraform will handle that with its state file. Or there may be cases where state file versioning may be required in decision-making. I would remove all the rollback logic from the cluster creation or update workflows.